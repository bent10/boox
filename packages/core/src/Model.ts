import { Encoder } from './Encoder.js'
import type {
  Dataset,
  Document,
  Documents,
  Features,
  ModelOptions
} from './types.js'

/**
 * Represents a model for full-text search using [TF-IDF](https://en.wikipedia.org/wiki/Tf-idf) score with
 * [inverted index](https://en.wikipedia.org/wiki/Inverted_index) weight.
 */
export class Model<T extends object = Dataset> {
  /**
   * The model encoder.
   */
  encoder: Encoder

  /**
   * The collection of documents indexed by reference identifier.
   */
  documents: Documents<T>

  /**
   * Represents the trained data as an inverted index, utilized for full-text
   * search. The index maps encoded information to relevant index entries,
   * facilitating structured database operations. Enables efficient retrieval of
   * information based on encoded representations.
   */
  features: Features

  /**
   * The dataset fields that will be indexed.
   */
  #featureKeys: string[]

  /**
   * Creates a new Model instance.
   *
   * @param options - Options for configuring the model.
   */
  constructor(options: ModelOptions = {}) {
    const { featureKeys = [], ...encoderOptions } = options

    this.encoder = new Encoder(encoderOptions)
    this.documents = {}
    this.features = {}

    this.#featureKeys = featureKeys
  }

  /**
   * Set of unique terms in the corpus.
   */
  get terms() {
    const terms = new Set<string>()
    for (const feature of Object.values(this.features)) {
      Object.keys(feature).forEach(term => terms.add(term))
    }
    return terms
  }

  /**
   * Trains the model with the provided document.
   *
   * @param document - The document to train the model with.
   */
  train(document: Document<T>) {
    const { id, attributes } = document
    // Store the magnitude for this document
    let magnitudeSquare = 0

    for (const key of this.#featureKeys) {
      const value = attributes[key as keyof T]

      if (value) {
        const normalizedValue = this.encoder.normalize(value)
        // encode feature value
        const encodings = this.encoder.encode(normalizedValue)

        if (!this.features[key]) {
          this.features[key] = {}
        }

        const attribute = this.features[key]

        encodings.forEach(encoding => {
          if (!attribute[encoding]) {
            attribute[encoding] = {}
          }

          const tf = this.calculateTermFrequency(encodings, encoding)
          attribute[encoding][id] = tf
          // update magnitude square
          magnitudeSquare += Math.pow(tf, 2)
        })
      }
    }

    // store the calculated magnitude for the document
    document.magnitude = Math.sqrt(magnitudeSquare)
    // store document
    this.documents[id] = document
  }

  /**
   * Updates the model with the provided document.
   *
   * @param document - The document to update the model with.
   */
  update(document: Document<T>) {
    // remove old document from features
    this.remove(document.id)
    // re-train
    this.train(document)
  }

  /**
   * Removes a document from the model.
   *
   * @param id - The identifier of the document to remove.
   */
  remove(id: string) {
    for (const key in this.features) {
      const attribute = this.features[key]

      for (const encoding in attribute) {
        const docs = attribute[encoding]

        for (const docId in docs) {
          if (docId === id) {
            delete docs[id]
          }
        }

        if (!Object.keys(docs).length) {
          delete attribute[encoding]
        }
      }
    }

    // remove document
    delete this.documents[id]
  }

  /**
   * Calculates the term frequency of an encoding token.
   *
   * @param encodings - The document encodings.
   * @param encoding - The encoding token.
   * @returns The term frequency of the encoding token.
   */
  calculateTermFrequency(encodings: string[], encoding: string): number {
    const termFrequency = encodings.reduce((count, te) => {
      return te === encoding ? count + 1 : count
    }, 0)

    // normalize term frequency by dividing by the total number of encodings
    return termFrequency / encodings.length
  }

  /**
   * Calculates the inverse document frequency of an encoding token.
   *
   * @param key - The feature key.
   * @param encoding - The encoding token.
   * @returns The inverse document frequency of the encoding token.
   */
  calculateIDF(encoding: string): number {
    const totalDocuments = Object.keys(this.documents).length
    let documentFrequency = 0

    for (const docId of Object.keys(this.documents)) {
      const hasDF = Object.values(this.features).some(
        attribute => !!attribute[encoding]?.[docId]
      )

      if (hasDF) {
        documentFrequency++
      }
    }

    // idf smooth
    return Math.log(totalDocuments / (1 + documentFrequency)) + 1
  }
}
