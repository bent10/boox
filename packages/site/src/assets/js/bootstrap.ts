import { Modal, Tooltip } from 'bootstrap'

// Constants
const TOOLTIP_SELECTOR = '[data-bs-toggle="tooltip"]'

/**
 * Initializes Bootstrap UI components.
 */
export function initBootstrapUI() {
  initializeTooltips()
}

/**
 * Initializes tooltip components.
 */
function initializeTooltips() {
  document.querySelectorAll(TOOLTIP_SELECTOR).forEach(el => {
    Tooltip.getOrCreateInstance(el)
  })
}

/**
 * Handles keyboard shortcuts for showing a modal.
 *
 * @param selector - The selector for the modal element.
 * @param key - The key to trigger the modal (default: '/').
 */
export function handleModalKeydown(selector: string, key = '/') {
  document.addEventListener('keydown', e => {
    if (e.key === key) {
      const modalElement = document.querySelector<HTMLElement>(selector)
      if (modalElement) {
        Modal.getOrCreateInstance(modalElement).show(modalElement)
      }
    }
  })
}
