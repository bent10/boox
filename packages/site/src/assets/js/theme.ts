// Constants
const THEME_TOGGLER_SELECTOR = '#theme-toggler'
const THEME_SWITCHER_SELECTOR = '[data-bs-theme-switcher]'
const DARK_THEME_MEDIA_QUERY = '(prefers-color-scheme: dark)'

/**
 * Handles theme switching functionality.
 *
 * @param selector - The selector for theme switcher buttons.
 */
export function handleThemeSwitcher(selector: string) {
  document.querySelectorAll<HTMLButtonElement>(selector).forEach(el => {
    el.addEventListener('click', () => {
      const themeName = el.dataset.bsThemeSwitcher || 'light'
      applyTheme(themeName, true)
    })
  })
}

/**
 * Sets up a listener for changes in the user's preferred color scheme.
 */
export function setupColorSchemeListener() {
  window
    .matchMedia(DARK_THEME_MEDIA_QUERY)
    .addEventListener('change', e => applyTheme(e.matches ? 'dark' : 'light'))
}

/**
 * Applies the specified theme and updates the UI.
 *
 * @param themeName - The name of the theme to apply (`'light'`, `'dark'`, or
 *   `'auto'`).
 * @param focusable - Whether to focus the theme toggler button after
 *   applying the theme.
 */
export function applyTheme(themeName: string, focusable = false) {
  const themeToggler = document.querySelector<HTMLButtonElement>(
    THEME_TOGGLER_SELECTOR
  )
  const themeSwitchers = document.querySelectorAll<HTMLElement>(
    THEME_SWITCHER_SELECTOR
  )

  themeSwitchers.forEach(el => {
    el.classList.toggle('active', el.dataset.bsThemeSwitcher === themeName)
    el.setAttribute(
      'aria-pressed',
      el.classList.contains('active') ? 'true' : 'false'
    )

    if (el.dataset.bsThemeSwitcher === themeName && themeToggler) {
      const icon = el.querySelector('svg')
      themeToggler.innerHTML = icon?.outerHTML || ''
    }
  })

  setTheme(themeName)
  setStoredTheme(themeName)

  if (focusable && themeToggler) {
    themeToggler.focus()
  }
}

/**
 * Gets the preferred theme based on user settings and system preferences.
 *
 * @returns The preferred theme name (`'light'`, `'dark'`, or
 *   `'auto'`).
 */
export function getPreferredTheme(): string {
  const storedTheme = getStoredTheme()
  if (storedTheme) {
    return storedTheme
  }

  return window.matchMedia(DARK_THEME_MEDIA_QUERY).matches ? 'dark' : 'light'
}

/**
 * Gets the stored theme from local storage.
 *
 * @returns The stored theme name or null if not found.
 */
function getStoredTheme(): string | null {
  return localStorage.getItem('theme')
}

/**
 * Sets the stored theme in local storage.
 *
 * @param theme - The theme name to store.
 */
function setStoredTheme(theme: string): void {
  localStorage.setItem('theme', theme)
}

/**
 * Sets the theme on the document element.
 *
 * @param theme - The theme name to set (`'light'`, `'dark'`, or
 *   `'auto'`).
 */
function setTheme(theme: string): void {
  if (theme === 'auto') {
    document.documentElement.setAttribute(
      'data-bs-theme',
      window.matchMedia(DARK_THEME_MEDIA_QUERY).matches ? 'dark' : 'light'
    )
  } else {
    document.documentElement.setAttribute('data-bs-theme', theme)
  }
}
