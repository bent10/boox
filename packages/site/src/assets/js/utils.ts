import type { State } from 'boox'

/**
 * Fetches data from the specified URL asynchronously and caches it until
 * the user closes the browser.
 *
 * @param url The URL from which to fetch the data.
 * @returns A Promise that resolves to the fetched data.
 * @throws
 *   Error if the network response was not OK or an error occurred during
 *   fetching.
 */
export async function fetchData<T extends object = object>(url: string) {
  try {
    const response = await fetch(url, {
      method: 'GET',
      mode: 'cors',
      headers: {
        'Content-Encoding': 'gzip',
        'Content-Type': 'application/json'
      },
      cache: 'default'
    })

    if (!response.ok) {
      throw new Error('Network response was not ok')
    }

    const blob = await response.blob()
    const ds = new DecompressionStream('gzip')
    const decompressedStream = blob.stream().pipeThrough(ds)

    return (await new Response(decompressedStream).json()) as State<T>
  } catch (error) {
    console.error(error)
    throw error
  }
}

/**
 * Creates a debounced version of a function that delays invoking the original
 * function until after `delay` milliseconds have elapsed since the last time
 * the debounced function was invoked.
 *
 * @param func - The function to debounce.
 * @param delay - The number of milliseconds to delay.
 * @returns A debounced function.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function debounce<F extends (...args: any[]) => any>(
  func: F,
  delay: number
): (...args: Parameters<F>) => Promise<ReturnType<F>> {
  let timeoutId: ReturnType<typeof setTimeout>

  return function debouncedFunction(
    this: ThisParameterType<F>,
    ...args: Parameters<F>
  ): Promise<ReturnType<F>> {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const context = this

    clearTimeout(timeoutId)

    return new Promise<ReturnType<F>>(resolve => {
      timeoutId = setTimeout(async () => {
        const result = await func.apply(context, args)
        resolve(result)
      }, delay)
    })
  }
}
