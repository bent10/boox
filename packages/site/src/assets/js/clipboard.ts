import { Tooltip } from 'bootstrap'
import ClipboardJS from 'clipboard'

const ICON_SUCCESS =
  '<svg xmlns="http://www.w3.org/2000/svg" width="1rem" height="1rem" viewBox="0 0 24 24"><g><path d="M22.641,4.232c-.427-.354-1.056-.296-1.409,.128L11.933,15.519l-.433-.433-1.286,1.543,1.078,1.078c.188,.188,.442,.293,.707,.293,.015,0,.03,0,.045,0,.281-.013,.543-.143,.724-.359L22.769,5.64c.354-.424,.296-1.055-.128-1.408Z" fill="var(--bs-success)"></path><path d="M6,18c-.265,0-.52-.105-.707-.293L1.293,13.707c-.391-.391-.391-1.023,0-1.414s1.023-.391,1.414,0l3.226,3.226L15.231,4.36c.354-.424,.983-.481,1.409-.128,.424,.354,.481,.984,.128,1.408L6.769,17.64c-.181,.216-.442,.346-.724,.359-.015,0-.03,0-.045,0Z" fill="currentColor"></path></g></svg>'

/**
 * Handles clipboard functionality for copying code snippets.
 */
export function handleClipboard() {
  // Initialize ClipboardJS
  const clipboard = new ClipboardJS('[data-js-copy]', {
    target: trigger =>
      trigger.parentElement?.parentElement?.querySelector('code') || trigger
  })

  clipboard.on('success', e => {
    e.clearSelection()

    const trigger = e.trigger as HTMLElement
    const icon = trigger.querySelector('svg')?.outerHTML || 'Copy'

    showSuccessIcon(trigger)
    setTimeout(() => resetTrigger(trigger, icon), 2000)
  })

  clipboard.on('error', e => {
    showFallbackMessage(e.trigger as HTMLElement)
  })
}

/**
 * Displays the success icon in the trigger element and updates the tooltip.
 *
 * @param trigger - The trigger element that initiated the copy action.
 */
function showSuccessIcon(trigger: HTMLElement) {
  trigger.innerHTML = ICON_SUCCESS
  trigger.classList.add('pe-none')

  const tooltip = Tooltip.getInstance(trigger)
  tooltip?.setContent({ '.tooltip-inner': 'Copied!' })
}

/**
 * Resets the trigger element to its original state and hides the tooltip.
 *
 * @param trigger - The trigger element that initiated the copy action.
 */
function resetTrigger(trigger: HTMLElement, icon: string) {
  trigger.innerHTML = icon
  trigger.classList.remove('pe-none')

  const tooltip = Tooltip.getInstance(trigger)
  tooltip?.setContent({ '.tooltip-inner': 'Copy to clipboard' })
  tooltip?.hide()
}

/**
 * Displays a fallback message in the tooltip for unsupported copy actions.
 *
 * @param trigger - The trigger element that initiated the copy action.
 */
function showFallbackMessage(trigger: HTMLElement) {
  const modifierKey = /mac/i.test(navigator.userAgent) ? '\u2318' : 'Ctrl-'
  const fallbackMsg = `Press ${modifierKey}C to copy`
  const tooltip = Tooltip.getInstance(trigger)

  tooltip?.setContent({ '.tooltip-inner': fallbackMsg })
  trigger.addEventListener(
    'hidden.bs.tooltip',
    () => tooltip?.setContent({ '.tooltip-inner': 'Copy to clipboard' }),
    { once: true }
  )
}
