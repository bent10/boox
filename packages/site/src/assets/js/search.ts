import { Modal } from 'bootstrap'
import Boox, { type SearchResultWithPagination } from 'boox'
import { doubleMetaphone } from 'double-metaphone'
import * as levenshtein from 'fastest-levenshtein'
import { stemmer } from 'stemmer'
import { removeStopwords } from 'stopword'
import { debounce, fetchData } from './utils.js'

/**
 * Documentation dataset object.
 */
export interface Dataset {
  id: string
  url: string
  title: string
  group: string
  section: string
  content: string
}

// Constants
const WORD_REGEX = /\b\w+\b/g
const RESULTS_PER_PAGE_MOBILE = 7
const RESULTS_PER_PAGE_DESKTOP = 10
const STATE_URL = 'https://stilearning.com/boox/docs-trained.gz?v=13'

// Boox instance
const boox = new Boox<Dataset>({
  /**
   * Configuration options for the Boox model.
   */
  modelOptions: {
    /**
     * Custom tokenizer function to split text into words and remove stopwords.
     *
     * @param input - The input text to tokenize.
     * @returns An array of tokens.
     */
    tokenizer(input) {
      const tokens = Array.from(input.match(WORD_REGEX) || [])
      return removeStopwords(tokens)
    },
    /**
     * Stemming function to reduce words to their root form.
     */
    stemmer: stemmer,
    /**
     * Phonetic encoding function using Double Metaphone.
     */
    phonetic: doubleMetaphone
  }
})

// Selectors for DOM elements
const mediaQueryList = window.matchMedia('(min-width: 576px)')
const searchModal = document.querySelector<HTMLInputElement>('#search-modal')!
const searchForm = document.querySelector<HTMLInputElement>('#search-form')!
const searchQuery = document.querySelector<HTMLInputElement>('#search-query')!
const searchResults =
  document.querySelector<HTMLInputElement>('#search-results')!
const resultsContainer = document.querySelector<HTMLInputElement>('#results')!
const resultPagination = document.querySelector<HTMLInputElement>(
  '#results-pagination'
)!

/**
 * Handles the search functionality, including fetching the search index state,
 * performing searches, and displaying results.
 */
export async function handleSearch() {
  // Fetch the search index state
  const currentState = await fetchData<Dataset>(STATE_URL)
  if (currentState) boox.currentState = currentState

  // Determine results per page based on screen size
  let resultsPerPage = mediaQueryList.matches
    ? RESULTS_PER_PAGE_DESKTOP
    : RESULTS_PER_PAGE_MOBILE

  // Event listeners for modal and media query changes
  searchModal.addEventListener('shown.bs.modal', () => searchQuery.focus())
  searchModal.addEventListener('hide.bs.modal', () => {
    searchQuery.value = ''
    searchQuery.dispatchEvent(new Event('input'))
  })
  mediaQueryList.addEventListener('change', (e: MediaQueryListEvent) => {
    resultsPerPage = e.matches
      ? RESULTS_PER_PAGE_DESKTOP
      : RESULTS_PER_PAGE_MOBILE
  })

  // Debounced search function triggered on input changes
  searchQuery.addEventListener('input', debounce(performSearch, 300))

  /**
   * Performs a search and updates the UI with the results.
   *
   * @param e - The input event object.
   */
  async function performSearch(e: Event) {
    const query = (e.target as HTMLInputElement).value

    const results = await boox.search(query, {
      limit: 70,
      highlightTag: ['<mark>', '</mark>'],
      useQueryVector: true,
      matchingCoefficient(query, attributes) {
        const { section } = attributes as unknown as Dataset

        const maxLength = Math.max(query.length, section.length)
        const distance = levenshtein.distance(query, section)
        const similarity = 1 - distance / maxLength

        return similarity
      }
    })

    const paginatedResults = Boox.paginateSearchResults<Dataset>(
      results,
      1,
      resultsPerPage
    )

    updateUIVisibility(!!paginatedResults.results.length)
    displayResults(paginatedResults)
  }
}

/**
 * Handles clicks on search results to hide a modal.
 *
 * @param resultsSelector - The selector for the search results container.
 * @param modalSelector - The selector for the modal element.
 */
export function handleClickedSearchResults(
  resultsSelector: string,
  modalSelector: string
) {
  document.querySelector(resultsSelector)?.addEventListener('click', e => {
    const target = e.target as HTMLElement
    if (isSearchResultItem(target)) {
      const modalElement = document.querySelector<HTMLElement>(modalSelector)
      if (modalElement) {
        Modal.getOrCreateInstance(modalElement).hide()
      }
    }
  })
}

/**
 * Checks if the given element is a search result item.
 *
 * @param element - The element to check.
 * @returns True if the element is a search result item, false otherwise.
 */
function isSearchResultItem(element: HTMLElement): boolean {
  return (
    element.classList.contains('list-group-item-action') ||
    !!element.parentElement?.classList.contains('list-group-item-action')
  )
}

/**
 * Displays the paginated search results in the UI.
 *
 * @param paginatedResults - The paginated search results.
 */
function displayResults(paginatedResults: SearchResultWithPagination<Dataset>) {
  resultsContainer.innerHTML = ''

  for (const result of paginatedResults.results) {
    const { url } = result.attributes
    const { text: title } = result.context('title')
    let { text: section } = result.context('section')
    const { text: kwic } = result.kwic('content', 2, true)[0] || {}

    if (!/<mark>/.test(section) && kwic) {
      section =
        '...<span class="d-inline-block overflow-hidden text-nowrap align-middle" style="max-width:80%;">' +
        kwic +
        '</span>...'
    }

    const resultElement = document.createElement('a')

    resultElement.classList.add('list-group-item', 'list-group-item-action')
    resultElement.href = url
    resultElement.innerHTML = `<h4 class="mb-1">${section}</h4>\n<p class="mb-0 opacity-75">${title}</p>`
    resultsContainer.appendChild(resultElement)
  }

  createPageResultsLinks(paginatedResults)
}

/**
 * Creates pagination links for the search results.
 *
 * @param paginatedResults - The paginated search results.
 */
function createPageResultsLinks(
  paginatedResults: SearchResultWithPagination<Dataset>
) {
  resultPagination.innerHTML = ''

  if (paginatedResults.totalPages < 2) return

  for (let i = 1; i <= paginatedResults.totalPages; i++) {
    const li = document.createElement('li')
    li.classList.add('page-item')
    if (i === paginatedResults.currentPage) {
      li.classList.add('active')
    }

    const a = document.createElement('a')
    a.classList.add('page-link', 'rounded', 'border-0')
    if (i < 10) a.classList.add('px-3')
    a.href = '#'
    a.textContent = String(i)
    a.addEventListener('click', e => {
      e.preventDefault()
      displayResults(paginatedResults.goTo(i))
    })

    li.appendChild(a)
    resultPagination.appendChild(li)
  }
}

/**
 * Updates the visibility of UI elements based on search results.
 *
 * @param hasResults - Whether there are any search results to display.
 */
function updateUIVisibility(hasResults: boolean) {
  if (hasResults) {
    searchForm.classList.remove('border-0')
    searchResults.classList.add('d-md-block')
    searchResults.classList.remove('d-md-none')
  } else {
    searchForm.classList.add('border-0')
    searchResults.classList.add('d-md-none')
    searchResults.classList.remove('d-md-block')
  }
}
