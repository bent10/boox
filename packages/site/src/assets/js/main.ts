import { handleModalKeydown, initBootstrapUI } from './bootstrap.js'
import { handleClipboard } from './clipboard.js'
import { handleSearch, handleClickedSearchResults } from './search.js'
import {
  applyTheme,
  getPreferredTheme,
  handleThemeSwitcher,
  setupColorSchemeListener
} from './theme.js'

initBootstrapUI()
handleModalKeydown('#search-modal', '/')
handleClickedSearchResults('#results', '#search-modal')

handleClipboard()

handleSearch()

applyTheme(getPreferredTheme())
handleThemeSwitcher('[data-bs-theme-switcher]')
setupColorSchemeListener()
