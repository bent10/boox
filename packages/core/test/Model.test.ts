/// <reference types="vitest/globals" />

import type { Document } from '../src/index.js'
import { Model } from '../src/Model.js'

interface Doc {
  title: string
  content: string
  [key: string]: unknown
}

describe('Model', () => {
  let model: Model<Doc>
  let document: Document<Doc>

  beforeEach(() => {
    document = {
      id: '1',
      attributes: {
        title: 'Foo Bar',
        content: 'This is a foo bar document.'
      },
      magnitude: 0
    }
    model = new Model<Doc>({
      featureKeys: ['title', 'content'],
      normalizer(input) {
        return input.trim()
      }
    })
  })

  it('should train the model with the provided document', () => {
    model.train(document)

    expect(model.features).toMatchSnapshot()
    expect(model.documents).toMatchSnapshot()
    expect(model.terms.has('foo')).toBe(true)
  })

  it('should normalize attribute value array', () => {
    model = new Model<Doc>({
      featureKeys: ['title', 'content', 'tags'],
      normalizer(input) {
        return input.trim()
      }
    })
    document.attributes.tags = ['foo', 'bar']
    model.train(document)

    expect(model.features).toMatchSnapshot()
    expect(model.documents).toMatchSnapshot()
    expect(model.terms.has('foo')).toBe(true)
  })

  it('should resolve unsupported feature format', () => {
    model = new Model<Doc>({
      featureKeys: ['title', 'content', 'tags', 'qux']
    })
    document.attributes.tags = ['foo', 'bar']
    document.attributes.qux = { qux: '' }
    model.train(document)

    expect(model.features).toMatchSnapshot()
    expect(model.documents).toMatchSnapshot()
    expect(model.terms.has('foo')).toBe(true)
  })

  it('should update the model with the provided document', () => {
    model.train(document)

    const updatedDocument: Document<Doc> = {
      id: '1',
      attributes: {
        title: 'Updated Title',
        content: 'This is an updated content.'
      },
      magnitude: 2
    }
    model.update(updatedDocument)

    expect(model.terms.has('foo')).toBe(false)
  })

  it('should remove a document from the model', () => {
    model.train(document)
    model.remove('1')
    const documents = model.documents

    expect(Object.keys(documents).length).toBe(0)
  })

  it('should calculate term frequency correctly', () => {
    const encodings = ['hello', 'world', 'hello', 'hello', 'world']
    const encoding = 'hello'
    const termFrequency = model.calculateTermFrequency(encodings, encoding)

    expect(termFrequency).toBe(0.6)
  })

  it('should calculate inverse document frequency correctly', () => {
    model.train(document)
    const idf = model.calculateIDF('foo')
    const missIDF = model.calculateIDF('hello')

    expect(idf).toBeGreaterThan(0) // 0.3068528194400547
    expect(missIDF).toBe(1)
  })
})
