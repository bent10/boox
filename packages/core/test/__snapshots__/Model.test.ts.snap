// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Model > should normalize attribute value array 1`] = `
{
  "content": {
    "a": {
      "1": 0.16666666666666666,
    },
    "bar": {
      "1": 0.16666666666666666,
    },
    "document": {
      "1": 0.16666666666666666,
    },
    "foo": {
      "1": 0.16666666666666666,
    },
    "is": {
      "1": 0.16666666666666666,
    },
    "this": {
      "1": 0.16666666666666666,
    },
  },
  "tags": {
    "bar": {
      "1": 0.5,
    },
    "foo": {
      "1": 0.5,
    },
  },
  "title": {
    "bar": {
      "1": 0.5,
    },
    "foo": {
      "1": 0.5,
    },
  },
}
`;

exports[`Model > should normalize attribute value array 2`] = `
{
  "1": {
    "attributes": {
      "content": "This is a foo bar document.",
      "tags": [
        "foo",
        "bar",
      ],
      "title": "Foo Bar",
    },
    "id": "1",
    "magnitude": 1.0801234497346435,
  },
}
`;

exports[`Model > should resolve unsupported feature format 1`] = `
{
  "content": {
    "a": {
      "1": 0.16666666666666666,
    },
    "bar": {
      "1": 0.16666666666666666,
    },
    "document": {
      "1": 0.16666666666666666,
    },
    "foo": {
      "1": 0.16666666666666666,
    },
    "is": {
      "1": 0.16666666666666666,
    },
    "this": {
      "1": 0.16666666666666666,
    },
  },
  "qux": {},
  "tags": {
    "bar": {
      "1": 0.5,
    },
    "foo": {
      "1": 0.5,
    },
  },
  "title": {
    "bar": {
      "1": 0.5,
    },
    "foo": {
      "1": 0.5,
    },
  },
}
`;

exports[`Model > should resolve unsupported feature format 2`] = `
{
  "1": {
    "attributes": {
      "content": "This is a foo bar document.",
      "qux": {
        "qux": "",
      },
      "tags": [
        "foo",
        "bar",
      ],
      "title": "Foo Bar",
    },
    "id": "1",
    "magnitude": 1.0801234497346435,
  },
}
`;

exports[`Model > should train the model with the provided document 1`] = `
{
  "content": {
    "a": {
      "1": 0.16666666666666666,
    },
    "bar": {
      "1": 0.16666666666666666,
    },
    "document": {
      "1": 0.16666666666666666,
    },
    "foo": {
      "1": 0.16666666666666666,
    },
    "is": {
      "1": 0.16666666666666666,
    },
    "this": {
      "1": 0.16666666666666666,
    },
  },
  "title": {
    "bar": {
      "1": 0.5,
    },
    "foo": {
      "1": 0.5,
    },
  },
}
`;

exports[`Model > should train the model with the provided document 2`] = `
{
  "1": {
    "attributes": {
      "content": "This is a foo bar document.",
      "title": "Foo Bar",
    },
    "id": "1",
    "magnitude": 0.816496580927726,
  },
}
`;
